misc/mempool.c

```diff
@@ -0,0 +1,346 @@
+			sys_dlist_append(&p->levels[l + 1].free_list, block2);
+		}
+	}
+
+	pool_irq_unlock(p, key);
+
+	return block;
+}
+
+int _sys_mem_pool_block_alloc(struct sys_mem_pool_base *p, size_t size,
+			      u32_t *level_p, u32_t *block_p, void **data_p)
+{
+	int i, from_l;
+	int alloc_l = -1, free_l = -1;
+	void *data;
+	size_t lsizes[p->n_levels];
+
+	/* Walk down through levels, finding the one from which we
+	 * want to allocate and the smallest one with a free entry
+	 * from which we can split an allocation if needed.  Along the
+	 * way, we populate an array of sizes for each level so we
+	 * don't need to waste RAM storing it.
+	 */
+	lsizes[0] = _ALIGN4(p->max_sz);
+	for (i = 0; i < p->n_levels; i++) {
+		if (i > 0) {
+			lsizes[i] = _ALIGN4(lsizes[i-1] / 4);
+		}
+
+		if (lsizes[i] < size) {
+			break;
+		}
+
+		alloc_l = i;
+		if (!level_empty(p, i)) {
+			free_l = i;
+		}
+	}
+
+	if (alloc_l < 0 || free_l < 0) {
+		*data_p = NULL;
+		return -ENOMEM;
+	}
+
+	/* Iteratively break the smallest enclosing block... */
+	data = block_alloc(p, free_l, lsizes[free_l]);
+
+	if (!data) {
+		/* This can happen if we race with another allocator.
+		 * It's OK, just back out and the timeout code will
+		 * retry.  Note mild overloading: -EAGAIN isn't for
+		 * propagation to the caller, it's to tell the loop in
+		 * k_mem_pool_alloc() to try again synchronously.  But
+		 * it means exactly what it says.
+		 *
+		 * This doesn't happen for user mode memory pools as this
+		 * entire function runs with a semaphore held.
+		 */
+		return -EAGAIN;
+	}
+
+	for (from_l = free_l; from_l < alloc_l; from_l++) {
+		data = block_break(p, data, from_l, lsizes);
+	}
+
+	*level_p = alloc_l;
+	*block_p = block_num(p, data, lsizes[alloc_l]);
+	*data_p = data;
+
+	return 0;
+}
+
+void _sys_mem_pool_block_free(struct sys_mem_pool_base *p, u32_t level,
+			      u32_t block)
+{
+	size_t lsizes[p->n_levels];
+	int i;
+
+	/* As in _sys_mem_pool_block_alloc(), we build a table of level sizes
+	 * to avoid having to store it in precious RAM bytes.
+	 * Overhead here is somewhat higher because block_free()
+	 * doesn't inherently need to traverse all the larger
+	 * sublevels.
+	 */
+	lsizes[0] = _ALIGN4(p->max_sz);
+	for (i = 1; i <= level; i++) {
+		lsizes[i] = _ALIGN4(lsizes[i-1] / 4);
+	}
+
+	block_free(p, level, lsizes, block);
+}
+
+
+/*
+ * Functions specific to user-mode blocks
+ */
+
+struct sys_mem_pool_block {
+	struct sys_mem_pool *pool;
+	u32_t block;
+	u32_t level;
+};
+
+void *sys_mem_pool_alloc(struct sys_mem_pool *p, size_t size)
+{
+	struct sys_mem_pool_block blk;
+	char *ret;
+
+	k_sem_take(p->sem, K_FOREVER);
+
+	size += sizeof(struct sys_mem_pool_block);
+	if (_sys_mem_pool_block_alloc(&p->base, size, &blk.level, &blk.block,
+				      (void **)&ret)) {
+		ret = NULL;
+		goto out;
+	}
+
+	memcpy(&blk, ret, sizeof(blk));
+	ret += sizeof(blk);
+out:
+	k_sem_give(p->sem);
+	return ret;
+}
+
+void sys_mem_pool_free(void *ptr)
```

Comment:
`free()` always work with NULL pointers (it takes in anything that `malloc()` can return).  It might be a good idea to do this as well, specially if we want this to be used by userland.